packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+import {
+  FetchHTTPClient,
+  HTTPClient,
+  HTTPClientRequest,
+  HTTPResponse,
+} from './http-client'
+import { SignOptions, sign } from 'jsonwebtoken'
+import { Emitter, backoff, sleep } from '@segment/analytics-core'
+import { AbortSignal, AbortController } from './abort'
+
+type AccessToken = {
+  access_token: string
+  expires_in: number
+}
+
+export type OAuthSettings = {
+  clientId: string
+  clientKey: Buffer
+  keyId: string
+  authServer?: string
+  scope?: string
+  maxRetries?: number
+  httpClient?: HTTPClient
+}
+
+export class TokenManager {
+  private alg = 'RS256' as const
+  private grantType = 'client_credentials' as const
+  private clientAssertionType =
+    'urn:ietf:params:oauth:client-assertion-type:jwt-bearer' as const
+  private clientId: string
+  private clientKey: Buffer
+  private keyId: string
+  private scope: string
+  private authServer: string
+  private httpClient: HTTPClient
+  private maxRetries: number
+  private clockSkewInSeconds = 0
+
+  private controller: AbortController
+  private signal: AbortSignal
+
+  private accessToken?: AccessToken
+  private tokenEmitter = new Emitter<{
+    access_token: [{ token: AccessToken } | { error: unknown }]
+  }>()
+  private retryCount: number
+  private lastError: any
+  private pollerTimer?: ReturnType<typeof setTimeout>
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
+    this.maxRetries = props.maxRetries ?? 3
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.controller = new AbortController()
+    this.signal = this.controller.signal
+    this.signal.addEventListener('abort', () => {
+      clearTimeout(this.pollerTimer)
+    })
+    this.retryCount = 0
+  }
+
+  async pollerLoop() {
+    let timeUntilRefreshInMs = 25
+    let response: HTTPResponse
+
+    try {
+      response = await this.requestAccessToken()
+    } catch (err) {
+      // Error without a status code - likely networking, retry
+      this.retryCount++
+      this.lastError = err
+
+      if (this.retryCount % this.maxRetries == 0) {
+        this.tokenEmitter.emit('access_token', { error: this.lastError })
+      }
+
+      this.pollerTimer = setTimeout(
+        this.pollerLoop.bind(this),
+        backoff({
+          attempt: this.retryCount,
+          minTimeout: 25,
+          maxTimeout: 1000,
+        })
+      ).unref()
+      return
+    }
+
+    if (response.headers !== undefined && response.headers.Date != undefined) {
+      const serverTime = Date.parse(response.headers.Date)
+      const skew = Date.now() - serverTime
+      if (this.clockSkewInSeconds == 0) {
```

Comment:
Nit: Prefer `===`

Though actually, could this whole thing be simplified as
`this.clockSkewInSeconds = (skew / 2)` since `0 / 2 === 0`?
I'm not sure why you would add the previous clock skew to the current calculated skew like you're doing in your else statement.

Reply:
For some reason I thought it was a good idea to take a sort of running average of the skew and the ===0 is the start condition.  It should probably just be `this.clockSkewInSeconds = Date.now() - servertime`, or if it's not completely unnecessary it could weigh the latest datapoint less than 50%...

Reply:
> take a sort of running average of the skew

Oh, I thought you were being fancy and dividing the skew by 2 to account for the fact there's latency between when the response generates the Date value and when we get it.